### ISSUE TITLE
Confusion about MultipleNegativesRankingLoss

### ISSUE THREAD

**Issue by i-wanna-to** [UKPLab/sentence-transformers#641]
First of all, thank you for your work.

I have been using sentence-transformers in recent days, but I am a bit confused about the MultipleNegativesRankingLoss.

If i use like below：

`from sentence_transformers import SentenceTransformer,  SentencesDataset, LoggingHandler, losses`
`from sentence_transformers.readers import InputExample`

`model = SentenceTransformer('distilbert-base-nli-mean-tokens')`
`train_examples = [InputExample(texts=['Anchor 1', 'Positive 1']),`
    `InputExample(texts=['Anchor 2', 'Positive 2'])]`

`train_dataset = SentencesDataset(train_examples, model)`
`train_dataloader = DataLoader(train_dataset, shuffle=True, batch_size=train_batch_size)`
`train_loss = losses.MultipleNegativesRankingLoss(model=model)`

`model.fit(train_objectives=[(train_dataloader, train_loss)], epochs=1, warmup_steps=100)`

Does it mean that **train_examples** are all positive examples? Will negative examples be constructed automatically by sentence_transformers?

Or, **train_examples** should also contain negative examples. Should I build them manually?

Thank you!

**Comment by nreimers** [comment]
The negatives are constructed automatically. All other "positives" in a batch are used as negative.

If you have a batch of 64:
anchor1, positive1
...
anchor64, positive64

Then you have (anchor1, positive1) as the positive and (positive2, ..., positive64) serve as negatives for this pair.

For the other pairs respectively.

**Comment by datistiquo** [comment]
I had the same confusion.:=) https://github.com/UKPLab/sentence-transformers/issues/608

So I would guess what you have to do, is using a dataset (generator) which prepares according to your specific batch size at the moment your data. So shifting each positive group away of size of batch size. A minimal code example how to start would be amazing @nreimers .:)

**Comment by datistiquo** [comment]
I think this is th edesired code example: https://github.com/UKPLab/sentence-transformers/issues/636
https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/datasets/sampler/LabelSampler.py

**Comment by i-wanna-to** [comment]
Okay, I understand. Thanks for your help! @nreimers @datistiquo

**Comment by datistiquo** [comment]
@nreimers Hey, I think it would be very helpful if you could provide at least a rough minimal example of using a dataset for MUltipleNegativeLoss structuring the batch right with having positive and negatives right, I think.
this loss is not applicable very well besides for very very large amount of data like in msmarco example. But in most situations you have to take care that all other positives are not negatives.

It is still a bit confusing. And I think an example would be great.
